### PR TITLE
LiteLLM tracing

### PR DESCRIPTION
**Pull request by B-Step62**
### What changes are proposed in this pull request?

Implement LiteLLM tracing integration. The [callback PR](https://github.com/BerriAI/litellm/pull/6147) on litellm side is complete and waiting for merging.

#### Challenge: Inline Display in Databricks Notebook

The callback setting logic is pretty straight-forward (just set it to global setting). However, the main complexity in this PR is handling **inline trace UI rendering in Databricks notebook**. The standard way to define tracing callback in LiteLLM is to implement two handlers; `log_success_event` and `log_failure_event`. MLflow callback also follows this, but the tricky thing is `log_success_event` is executed in a background thread and not waited ([code](https://github.com/BerriAI/litellm/blob/4f8a3fd4cfc20cf43b38379928b41c2691c85d36/litellm/utils.py#L7797)) (the failure handler is simply run synchronously).

Why this causes an issue in DB notebook? This is because inline trace UI requires all traces to be generated before rendering the cell output. In other words, trace logging should be completed before the cell execution finishes. However, this is not always true for `log_success_event` case due to that async design. If a trace is generated after cell finish, it will be displayed in a random different cell executed later - which is pretty bad UX.

#### Solution 1: Make logging sync (not accepted by LiteLLM)

The first solution we came up was adding an option to call `log_success_event` handler synchronously, similarly to the failure handler. We propose adding special handling for logger that requires sync logging (like mlflow). However, LiteLLM maintainers did not prefer this approach (which is understandable).

#### Solution 2: Patch thread (applied in this PR)

Second option is monkey-patching their API to wait for trace generation. This involves moderately complex patching logic because they don't expose the thread spawning logic as a dedicated API. It is not too complex (about 100 lines long) so hopefully it doesn't add too much maintenance burden.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (WIP)

In Databricks:

Screenshot 2024-11-06 at 15 54 44

Screenshot 2024-11-06 at 15 55 02

In OSS:

Screenshot 2024-11-06 at 16 02 04

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

Will add to the tracing guide once the UI sync is done this week, so we can include the UI updates.

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

MLflow Tracing added integration for LiteLLM. Now you can enable auto-tracing for LiteLLM by simply calling `mlflow.litellm.autolog()`.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/ml-package-versions.yml [hunk 1]
```diff
@@ -879,3 +879,15 @@ promptflow:
...
     run: |
       pytest tests/promptflow/test_promptflow_model_export.py
+
+litellm:
+  package_info:
+    pip_release: "litellm"
+    install_dev: |
+      pip install git+https://github.com/BerriAI/litellm.git
+  autologging:
+    minimum: "1.52.0"
```
Comment (B-Step62) [inline mlflow/ml-package-versions.yml:889]: self-TODO: Update minimum version once the callback PR in LiteLLM side is merged.

File: mlflow/litellm/__init__.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        # - https://github.com/BerriAI/litellm/blob/4f8a3fd4cfc20cf43b38379928b41c2691c85d36/litellm/utils.py#L946
+        # - https://github.com/BerriAI/litellm/blob/4f8a3fd4cfc20cf43b38379928b41c2691c85d36/litellm/utils.py#L7526
+        if target and target.__name__ in [
+            "success_handler",
+            "run_success_logging_and_cache_storage",
+        ]:
+            logging_threads.append(self)
+        return original(self, *args, **kwargs)
+
+    Thread.start = patched_thread
```
Comment (harupy) [inline mlflow/litellm/__init__.py:128]: This looks a bit scary. Is this the only way to accomplish what we want?

Comment (B-Step62) [reply mlflow/litellm/__init__.py:128]: This is ugly but I could not find a better way to ensure that random thread to finish. Definitely happy to change if you have an idea😂

File: mlflow/litellm/__init__.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                _patch_theading_in_function(litellm, func)
+
+            # For streaming case, we need to patch the iterator because traces are generated
+            # when consuming the generator, not when calling the main APIs.
+            _patch_theading_in_function(litellm.utils.CustomStreamWrapper, "__next__")
+
+            # NB: We don't need to patch async function because Databricks notebook waits
+            # for the async task to finish before finishing the cell.
+    else:
+        litellm.success_callback = _remove_mlflow_callbacks(litellm.success_callback)
```
Comment (serena-ruan) [inline mlflow/litellm/__init__.py:68]: Should we remove mlflow callbacks if it's added by the user?

Comment (serena-ruan) [reply mlflow/litellm/__init__.py:68]: I think this question matches your comment on `_autolog` about `This prevents us from removing our callback and patching when autologging is turned off.` But I though autologging means, if the model doesn't contain any callback then we inject them, if it already contains we do nothing (same for langchain)

File: mlflow/litellm/__init__.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    Thread.start = patched_thread
+    try:
+        yield logging_threads
+    finally:
+        Thread.start = original
+
+
+def _append_mlflow_callbacks(original_callbacks):
+    if not any(cb == "mlflow" for cb in original_callbacks):
```
Comment (serena-ruan) [inline mlflow/litellm/__init__.py:136]: Is it possible original_callbacks is None?

File: mlflow/litellm/__init__.py (hunk 3 above)
Comment (B-Step62) [reply mlflow/litellm/__init__.py:68]: > Should we remove mlflow callbacks if it's added by the user?

Technically not, but in practice, I wonder how often users will inject MLflow callback by themselves - imo it's near zero. We could have a global state to determine if the callback is injected by MLflow or not, but I double it gives more benefit than its complexity🤔

File: mlflow/litellm/__init__.py (hunk 4 above)
Comment (B-Step62) [reply mlflow/litellm/__init__.py:136]: Not really - they are initialized with an empty list: https://github.com/BerriAI/litellm/blob/70aa85af1fc01d30934ee71a703b8d3982420d05/litellm/__init__.py#L39-L42

File: mlflow/litellm/__init__.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/litellm/__init__.py:68]: Oh the callback is global in liteLLM? I thought it's defined within each model call.... Emm if so then seems we have to remove mlflow callback when disable=True..

Comment (B-Step62) [reply mlflow/litellm/__init__.py:68]: Right, it is global callback similar to LlamaIndex, unlike LangChain

**Comment by B-Step62** [comment]
@harupy @serena-ruan Would you mind taking another look? Code change in LiteLLM side is released.

**Review by serena-ruan — APPROVED** [review]
LGTM!

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**